### PR TITLE
test: Load new model version should not reload loaded existing model version(s)

### PR DESCRIPTION
**Pull request by kthui**
#### What does the PR do?

Add the following tests on model version reload:
1. If a version is loaded and unmodified, then it should not be reloaded on the next load request.
2. If a version is loaded but modified, then it should be reloaded on the next load request.
3. If a version is not loaded and its model file is in the model directory, then it should be loaded on the next load request.
4. If a version is not loaded nor its model file is in the model directory, then it should be loaded on the next load request.
5. If a generic file is modified on the model directory, i.e. `model_directory/common_dependency.py`, then all loaded version(s) should be reloaded on the next load request.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [x] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [x] test

#### Related PRs:

https://github.com/triton-inference-server/core/pull/388

#### Where should the reviewer start?

Start with the core PR.

#### Test plan:

The PR adds the test, see the Test plan on the core PR.

- CI Pipeline ID: 17511909  17650341

#### Caveats:

N/A

#### Background

N/A

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
N/A

### CODE REVIEW

**Review by GuanLuo — APPROVED** [review]

File: qa/L0_lifecycle/lifecycle_test.py [hunk 1]
```diff
@@ -3493,6 +3493,79 @@ def test_delete_custom_config(self):
             except Exception as ex:
                 self.assertTrue(False, "unexpected error {}".format(ex))
 
+    def test_load_new_model_version(self):
```
Comment (rmccorm4) [inline qa/L0_lifecycle/lifecycle_test.py:3496]: Did you ever test the scenario where we change the `version_policy` in a way that we expect to trigger certain loads, to make sure "ignoring" the version policy in the core PR doesn't break that?

**Review by rmccorm4 — COMMENTED** [review]
Non-blocking for this release, but curious about this comment: https://github.com/triton-inference-server/server/pull/7527/files#r1722556601

File: qa/L0_lifecycle/lifecycle_test.py (hunk 1 above)
Comment (kthui) [reply qa/L0_lifecycle/lifecycle_test.py:3496]: Added the test: [Expand test to cover unspecifying loaded versions](https://github.com/triton-inference-server/server/pull/7527/commits/6beab1c88382992b672b611df2ae027d6620321e)

Comment (kthui) [reply qa/L0_lifecycle/lifecycle_test.py:3496]: It is working locally, waiting on CI to rerun the L0_lifecycle with updated test script.

File: qa/L0_lifecycle/lifecycle_test.py [hunk 2]
```diff
@@ -3566,6 +3566,33 @@ def test_load_new_model_version(self):
         self.assertEqual(server_log.count("[PB model] Loading version 4"), 2)
         self.assertEqual(server_log.count("successfully loaded 'identity_fp32'"), 3)
 
+        # update model config to only load version 4
+        config_path = os.path.join("models", model_name, "config.pbtxt")
+        with open(config_path, mode="r+", encoding="utf-8", errors="strict") as f:
+            config = f.read()
+            config = config.replace(
+                "version_policy: { specific: { versions: [1, 2, 3, 4] } }",
+                "version_policy: { specific: { versions: [4] } }",
```
Comment (rmccorm4) [inline qa/L0_lifecycle/lifecycle_test.py:3575]: I think this test case is confirming that upon changing version_policy to just version 4 (which is already loaded), no reload is triggered. 

But I think the test case I'm describing is the opposite -- if you change _only_ the version_policy to a version that isn't already loaded - will it load the model, or will the version policy change be ignored when considering reload criteria in core?

Comment (rmccorm4) [reply qa/L0_lifecycle/lifecycle_test.py:3575]: Does that make sense?

Comment (kthui) [reply qa/L0_lifecycle/lifecycle_test.py:3575]: I see, a complete load test with only version policy change, added [Expand test to cover load action with only version addition to policy](https://github.com/triton-inference-server/server/pull/7527/commits/3c6980501dd66ee3a575050987e56372d7f5d71a)

Comment (kthui) [reply qa/L0_lifecycle/lifecycle_test.py:3575]: It is working locally, waiting on CI to rerun the L0_lifecycle with updated test script.

**Review by GuanLuo — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]

File: qa/L0_lifecycle/lifecycle_test.py (hunk 2 above)
Comment (rmccorm4) [reply qa/L0_lifecycle/lifecycle_test.py:3575]: Nice, thanks!

**Pull request merged**